### PR TITLE
User sotry finished

### PR DESCRIPTION
**Pull request by HaojunLi**
Hellor Robert, I finished the user story. Exception works well but I cannot create database because I did not have permission. Maybe you can test it. If you meet any problem, please let me know. I also fix some sql injection problems and our code is now safe. Thank you.

### CODE REVIEW

File: WebUI/WebUI/middlewear/database.py [hunk 1]
```diff
@@ -15,9 +15,21 @@ def __init__(self, get_response):
         self.get_response = get_response
         self.refresh()
 
+    # calls after __call__
+    # I modify here because __call__ is called before process_view, but we need to do refresh before other operations
+    # so after refresh, we repeat the operation.
+    # This function resposne to index function, the function of our home page.
+    def process_view(self,request, index, view_args, view_kwargs):
```
Comment (rjzupkoii) [inline WebUI/WebUI/middlewear/database.py:22]: This seems a bit redundant given the code in __call__. Can just process_view be used by itself?

File: WebUI/lims/AppDatabase.py [hunk 2]
```diff
@@ -28,13 +29,21 @@ class AppDatabase:
                    ALTER TABLE public.v_replicates OWNER TO sim;"""
 
 
-    def cloneDatabase(self, username, password, database):
+    def cloneDatabase(self, request, username, password, database):
```
Comment (rjzupkoii) [inline WebUI/lims/AppDatabase.py:32]: What information are you getting from request?

File: WebUI/lims/shared.py [hunk 3]
```diff
@@ -74,3 +74,19 @@ def visitor_ip_address(request):
     else:
         ip = request.META.get('REMOTE_ADDR')
     return ip
+
+'''
```
Comment (rjzupkoii) [inline WebUI/lims/shared.py]: These are all standard conventions that developers will know, so they shouldn't be included in the function documentation.

File: WebUI/lims/shared.py [hunk 4]
```diff
@@ -74,3 +74,19 @@ def visitor_ip_address(request):
...
+^	Starts with	
+$	Ends with	
+*	Zero or more occurrences
++	One or more occurrences
+{}	Exactly the specified number of occurrences
+|	Either or	
+()	Capture and group
+'''
+def regCheck(regularExp,stringIn):
+    result = re.search(regularExp, stringIn)
```
Comment (rjzupkoii) [inline WebUI/lims/shared.py]: It doesn't seem like this function is really doing much that just directly calling re.search is. I'd recommend removing the function and just directly calling re.search in the conditional.

File: WebUI/lims/views.py [hunk 5]
```diff
@@ -255,15 +258,25 @@ def DeleteNotes(request, studyId, id):
 def createDatabase(request):
 
     # Prepare the connection string
-    username = ''    # TODO get the username from the form
-    password = ''    # TODO Get password from the form
+    username = request.POST['userName']    # TODO get the username from the form
```
Comment (rjzupkoii) [inline WebUI/lims/views.py]: Once the task is done you can remove the TODO

File: WebUI/lims/views.py [hunk 6]
```diff
@@ -255,15 +258,25 @@ def DeleteNotes(request, studyId, id):
...
     
     # Prepare the query
-    database = ''    # TODO Get database from form    
-    
+    database = request.POST['databaseName']    # TODO Get database from form
+    # Database name should be checked for compliance with SQL lexicon before the operation takes place
+    # Effectively, 1 to 31 characters, start with letter or underscore (_) may contain letters, numbers, or underscores after that.
+    result = regCheck("^[A-Za-z_][A-Za-z\d_]{0,31}$",database)
+    # If database name does not satisfy the requirements
+    if not result:
```
Comment (rjzupkoii) [inline WebUI/lims/views.py]: See comment above, but just make the call to re as part of the if conditional.

**Review by rjzupkoii — CHANGES_REQUESTED** [review]

File: WebUI/WebUI/middlewear/database.py (hunk 1 above)
Comment (HaojunLi) [reply WebUI/WebUI/middlewear/database.py:22]: process_view only response to one specific function in view.py. i.e. this process_view only response to index function. However, __call__ response to all functions in view.py.  __call__ is called before process_view, so, inside process_view, we need to call functions again to overwrite previous steps.

File: WebUI/lims/AppDatabase.py (hunk 2 above)
Comment (HaojunLi) [reply WebUI/lims/AppDatabase.py:32]: commitQuery inside this function needs parameter request.

File: WebUI/lims/shared.py (hunk 3 above)
Comment (HaojunLi) [reply WebUI/lims/shared.py]: I will delete them and add new comments.

File: WebUI/lims/shared.py (hunk 4 above)
Comment (HaojunLi) [reply WebUI/lims/shared.py]: Ok, I will change this.

File: WebUI/lims/views.py (hunk 5 above)
Comment (HaojunLi) [reply WebUI/lims/views.py]: I will remove them

File: WebUI/lims/views.py (hunk 6 above)
Comment (HaojunLi) [reply WebUI/lims/views.py]: No problem

**Pull request closed without merging**